// @airgap/serializer/v3/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IACMessageType = void 0;
var IACMessageType;
(function (IACMessageType) {
    // PairRequest = 1,
    // PairResponse = 2,
    IACMessageType[IACMessageType["AccountShareRequest"] = 3] = "AccountShareRequest";
    IACMessageType[IACMessageType["AccountShareResponse"] = 4] = "AccountShareResponse";
    IACMessageType[IACMessageType["TransactionSignRequest"] = 5] = "TransactionSignRequest";
    IACMessageType[IACMessageType["TransactionSignResponse"] = 6] = "TransactionSignResponse";
    IACMessageType[IACMessageType["MessageSignRequest"] = 7] = "MessageSignRequest";
    IACMessageType[IACMessageType["MessageSignResponse"] = 8] = "MessageSignResponse";
    // MessageVerifyRequest = 9,
    // MessageVerifyResponse = 10,
    // MessageEncryptRequest = 11,
    // MessageEncryptResponse = 12,
    // MessageDecryptRequest = 13,
    // MessageDecryptResponse = 14,
    // ItemStoreRequest = 15,
    // ItemStoreResponse = 16,
    // ItemRetrieveRequest = 17,
    // ItemRetrieveResponse = 18,
    // ConfigSetRequest = 19,
    // ConfigSetResponse = 20,
    // MultisigRequest = 21,
    // MultisigResponse = 22,
    // SocialRecoveryShareRequest = 23,
    // SocialRecoveryShareResponse = 24
})(IACMessageType = exports.IACMessageType || (exports.IACMessageType = {}));
