// @airgap/coinlib-core/utils/ProtocolSymbols.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isProtocolSymbol = exports.isSubProtocolSymbol = exports.isMainProtocolSymbol = exports.SubProtocolSymbols = exports.MainProtocolSymbols = void 0;
var MainProtocolSymbols;
(function (MainProtocolSymbols) {
    MainProtocolSymbols["ACURAST"] = "acurast";
    MainProtocolSymbols["AE"] = "ae";
    MainProtocolSymbols["BTC"] = "btc";
    MainProtocolSymbols["BTC_SEGWIT"] = "btc_segwit";
    MainProtocolSymbols["ETH"] = "eth";
    MainProtocolSymbols["XTZ"] = "xtz";
    MainProtocolSymbols["XTZ_SHIELDED"] = "xtz_shielded";
    MainProtocolSymbols["GRS"] = "grs";
    MainProtocolSymbols["COSMOS"] = "cosmos";
    MainProtocolSymbols["POLKADOT"] = "polkadot";
    MainProtocolSymbols["KUSAMA"] = "kusama";
    MainProtocolSymbols["MOONBASE"] = "moonbase";
    MainProtocolSymbols["MOONRIVER"] = "moonriver";
    MainProtocolSymbols["MOONBEAM"] = "moonbeam";
    MainProtocolSymbols["ASTAR"] = "astar";
    MainProtocolSymbols["SHIDEN"] = "shiden";
    MainProtocolSymbols["ICP"] = "icp";
    MainProtocolSymbols["ICP_CKBTC"] = "icp_ckbtc";
    MainProtocolSymbols["COREUM"] = "coreum";
    MainProtocolSymbols["OPTIMISM"] = "optimism";
    MainProtocolSymbols["MINA"] = "mina";
    MainProtocolSymbols["STELLAR"] = "stellar";
})(MainProtocolSymbols = exports.MainProtocolSymbols || (exports.MainProtocolSymbols = {}));
var SubProtocolSymbols;
(function (SubProtocolSymbols) {
    SubProtocolSymbols["XTZ_KT"] = "xtz-kt";
    SubProtocolSymbols["XTZ_BTC"] = "xtz-btc";
    SubProtocolSymbols["XTZ_USD"] = "xtz-usd";
    SubProtocolSymbols["XTZ_KUSD"] = "xtz-kusd";
    SubProtocolSymbols["XTZ_STKR"] = "xtz-stkr";
    SubProtocolSymbols["XTZ_ETHTZ"] = "xtz-eth";
    SubProtocolSymbols["XTZ_UUSD"] = "xtz-uusd";
    SubProtocolSymbols["XTZ_YOU"] = "xtz-you";
    SubProtocolSymbols["XTZ_W"] = "xtz-w";
    SubProtocolSymbols["XTZ_UDEFI"] = "xtz-udefi";
    SubProtocolSymbols["XTZ_UBTC"] = "xtz-ubtc";
    SubProtocolSymbols["XTZ_UXTZ"] = "xtz-uxtz";
    SubProtocolSymbols["XTZ_CTEZ"] = "xtz-ctez";
    SubProtocolSymbols["XTZ_PLENTY"] = "xtz-plenty";
    SubProtocolSymbols["XTZ_WRAP"] = "xtz-wrap";
    SubProtocolSymbols["XTZ_QUIPU"] = "xtz-quipu";
    SubProtocolSymbols["XTZ_DOGA"] = "xtz-doga";
    SubProtocolSymbols["XTZ_BTC_TEZ"] = "xtz-btc-tez";
    SubProtocolSymbols["XTZ_USDT"] = "xtz-usdt";
    SubProtocolSymbols["XTZ_SIRS"] = "xtz-sirs";
    SubProtocolSymbols["ETH_ERC20"] = "eth-erc20";
    SubProtocolSymbols["ETH_ERC20_XCHF"] = "eth-erc20-xchf";
    SubProtocolSymbols["OPTIMISM_ERC20"] = "optimism-erc20";
    SubProtocolSymbols["STELLAR_ASSET"] = "stellar-asset";
})(SubProtocolSymbols = exports.SubProtocolSymbols || (exports.SubProtocolSymbols = {}));
function isMainProtocolSymbol(identifier) {
    return Object.values(MainProtocolSymbols).includes(identifier);
}
exports.isMainProtocolSymbol = isMainProtocolSymbol;
function isSubProtocolSymbol(identifier) {
    return Object.values(SubProtocolSymbols).includes(identifier);
}
exports.isSubProtocolSymbol = isSubProtocolSymbol;
function isProtocolSymbol(identifier) {
    return isMainProtocolSymbol(identifier) || isSubProtocolSymbol(identifier);
}
exports.isProtocolSymbol = isProtocolSymbol;
