// @airgap/coinlib-core/interfaces/IAirGapTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AirGapTransactionWarningType = exports.AirGapTransactionStatus = exports.AirGapTransactionType = void 0;
var AirGapTransactionType;
(function (AirGapTransactionType) {
    AirGapTransactionType["SPEND"] = "Spend Transaction";
    AirGapTransactionType["DELEGATE"] = "Delegation";
    AirGapTransactionType["UNDELEGATE"] = "Undelegate";
})(AirGapTransactionType = exports.AirGapTransactionType || (exports.AirGapTransactionType = {}));
var AirGapTransactionStatus;
(function (AirGapTransactionStatus) {
    AirGapTransactionStatus["APPLIED"] = "applied";
    AirGapTransactionStatus["FAILED"] = "failed";
})(AirGapTransactionStatus = exports.AirGapTransactionStatus || (exports.AirGapTransactionStatus = {}));
var AirGapTransactionWarningType;
(function (AirGapTransactionWarningType) {
    AirGapTransactionWarningType["SUCCESS"] = "success";
    AirGapTransactionWarningType["NOTE"] = "note";
    AirGapTransactionWarningType["WARNING"] = "warning";
    AirGapTransactionWarningType["ERROR"] = "error";
})(AirGapTransactionWarningType = exports.AirGapTransactionWarningType || (exports.AirGapTransactionWarningType = {}));
