// @airgap/serializer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateSyncSchemeV2 = exports.validateSyncScheme = exports.generateIdV2 = exports.generateId = exports.Message = exports.SerializerV3 = exports.Serializer = exports.IACMessageType = void 0;
var interfaces_1 = require("./v2/interfaces");
Object.defineProperty(exports, "IACMessageType", { enumerable: true, get: function () { return interfaces_1.IACMessageType; } });
var message_1 = require("./v2/message");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return message_1.Message; } });
var serializer_1 = require("./v2/serializer");
Object.defineProperty(exports, "Serializer", { enumerable: true, get: function () { return serializer_1.Serializer; } });
var generateId_1 = require("./v2/utils/generateId");
Object.defineProperty(exports, "generateIdV2", { enumerable: true, get: function () { return generateId_1.generateIdV2; } });
var validators_1 = require("./v2/validators/validators");
Object.defineProperty(exports, "validateSyncSchemeV2", { enumerable: true, get: function () { return validators_1.validateSyncScheme; } });
var serializer_2 = require("./v3/serializer");
Object.defineProperty(exports, "SerializerV3", { enumerable: true, get: function () { return serializer_2.SerializerV3; } });
var generateId_2 = require("./v3/utils/generateId");
Object.defineProperty(exports, "generateId", { enumerable: true, get: function () { return generateId_2.generateId; } });
var validators_2 = require("./v3/validators/validators");
Object.defineProperty(exports, "validateSyncScheme", { enumerable: true, get: function () { return validators_2.validateSyncScheme; } });
